### PR TITLE
Check the provided password against all albums

### LINKED ISSUE #724: [Enhancement] - heirarchical password check for password protected albums
At present - if you use password protection together with sub-albums then you can either:

Set the password on just the parent - in this case - you can see the thumbs of the sub-albums when you are showing the parent and can navigate in - but - if you head to the url of a sub-album directly then its open

Set the password on the parent and all sub-albums - in this case - the sub-albums are not open via direct link - but - if you start at the top and work down - you can only see padlock icons and for each new sub-album you have to enter the password again.

It would be nice to be able to check up the tree so that a sub-album was protected by its parent's (or higher in the tree) settings.

### PR DESCRIPTION
**Pull request by kamil4**
Fixes #724

### CODE REVIEW

File: app/Http/Controllers/AlbumController.php [hunk 1]
```diff
@@ -251,8 +251,21 @@ public function getPublic(Request $request)
 					if ($this->sessionFunctions->has_visible_album($album->id)) {
 						return 'true';
 					}
-					if ($album->password == '' || Hash::check($request['password'], $album->password)) {
```
Comment (d7415) [inline app/Http/Controllers/AlbumController.php:254]: I take it the `$album->password == '' || ` part is handled elsewhere so not required here?

Comment (kamil4) [reply app/Http/Controllers/AlbumController.php:254]: Yeah, exactly six lines earlier...

**Review by d7415 — APPROVED** [review]
Makes sense to me.

File: app/Http/Controllers/AlbumController.php [hunk 2]
```diff
@@ -251,8 +251,21 @@ public function getPublic(Request $request)
...
-						$this->sessionFunctions->add_visible_album($album->id);
+					if (Hash::check($request['password'], $album->password)) {
+						// We add all the albums that the password unlocks so
+						// that the user is not repeatedly asked to enter the
+						// password as they browse through the hierarchy.  This
+						// should be safe as the list of such albums is not
+						// exposed to the user and is considered as the last
+						// access check criteria.
+						$albums = Album::get();
```
Comment (ildyria) [inline app/Http/Controllers/AlbumController.php]: This should be refined with a `where('password','!=','')`

File: app/Http/Controllers/AlbumController.php [hunk 3]
```diff
@@ -251,8 +251,21 @@ public function getPublic(Request $request)
...
+						$albumIDs = [];
+						foreach ($albums as $album) {
+							if ($album->password !== '' && Hash::check($request['password'], $album->password)) {
```
Comment (ildyria) [inline app/Http/Controllers/AlbumController.php]: `$album->password !== ''` is now useless once the where condition is applied to the Query.

**Review by ildyria — CHANGES_REQUESTED** [review]
some refinement to avoid looping over ALL albums.

File: app/Http/Controllers/AlbumController.php (hunk 2 above)
Comment (kamil4) [reply app/Http/Controllers/AlbumController.php]: Good point, I don't know what I was thinking when I wrote this...

**Comment by kamil4** [comment]
Should be good to go now...

File: app/Http/Controllers/AlbumController.php [hunk 4]
```diff
@@ -251,8 +251,21 @@ public function getPublic(Request $request)
...
-					if ($album->password == '' || Hash::check($request['password'], $album->password)) {
-						$this->sessionFunctions->add_visible_album($album->id);
+					if (Hash::check($request['password'], $album->password)) {
+						// We add all the albums that the password unlocks so
+						// that the user is not repeatedly asked to enter the
+						// password as they browse through the hierarchy.  This
+						// should be safe as the list of such albums is not
+						// exposed to the user and is considered as the last
+						// access check criteria.
+						$albums = Album::whereNotNull('password')->get();
```
Comment (ildyria) [inline app/Http/Controllers/AlbumController.php]:
```suggestion
						$albums = Album::whereNotNull('password')->where('password', '!=', '')->get();
```

Comment (ildyria) [reply app/Http/Controllers/AlbumController.php]: just to be sure 😄

**Pull request merged**